source.go

```diff
@@ -43,18 +56,215 @@ func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
+	if !block.GetHasChildren() {
+		sdk.Logger(ctx).Debug().
+			Str("block_id", block.GetID().String()).
+			Msg("block has no children")
+		return children, nil
+	}
+
+	fetch := true
+	var cursor notion.Cursor
+	for fetch {
+		resp, err := s.client.Block.GetChildren(
+			ctx,
+			block.GetID(),
+			&notion.Pagination{
+				StartCursor: cursor,
+			},
+		)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed getting children for block ID %v, cursor %v: %w",
+				block.GetID(),
+				cursor,
+				err,
+			)
+		}
+
+		// get grandchildren as well
+		for _, child := range resp.Results {
+			children = append(children, child)
+			grandChildren, err := s.getChildren(ctx, child)
+			if err != nil {
+				return nil, err
+			}
+			children = append(children, grandChildren...)
+		}
+
+		fetch = resp.HasMore
+		cursor = notion.Cursor(resp.NextCursor)
+	}
+	return children, nil
 }
 
-func (s *Source) Ack(ctx context.Context, position sdk.Position) error {
+func (s *Source) Ack(context.Context, sdk.Position) error {
 	return nil
 }
 
-func (s *Source) Teardown(ctx context.Context) error {
+func (s *Source) Teardown(context.Context) error {
 	return nil
 }
+
+func (s *Source) populateIDs(ctx context.Context) error {
+	if len(s.fetchIDs) > 0 {
+		return nil
+	}
+	// the first read attempt (when the connector starts)
+	if !s.firstFetch {
+		sdk.Logger(ctx).Debug().
+			Dur("poll_interval", s.config.pollInterval).
+			Msg("sleeping before checking for changes")
+		time.Sleep(s.config.pollInterval)
+	}
+	s.firstFetch = false
+
+	sdk.Logger(ctx).Debug().Msg("populating IDs")
+	fetch := true
+	var cursor notion.Cursor
+	for fetch {
+		results, err := s.getPages(ctx, cursor)
+		if err != nil {
+			return fmt.Errorf("search failed: %w", err)
+		}
+		s.addToFetchIDs(ctx, results)
+
+		fetch = results.HasMore
+		cursor = results.NextCursor
+	}
+
+	sdk.Logger(ctx).Info().Msgf("fetched %v IDs", len(s.fetchIDs))
+	return nil
+}
+
+func (s *Source) addToFetchIDs(ctx context.Context, results *notion.SearchResponse) {
+	for _, result := range results.Results {
+		switch result.GetObject().String() {
+		case "page":
+			page := result.(*notion.Page)
+			sdk.Logger(ctx).Trace().
+				Str("page_id", page.ID.String()).
+				Time("last_edited_time", page.LastEditedTime).
+				Time("created_time", page.CreatedTime).
+				Msg("checking if page has changed")
+			if s.hasChanged(page) {
+				s.fetchIDs = append(s.fetchIDs, page.ID.String())
+			}
+		default:
+			sdk.Logger(ctx).Warn().
+				Str("object_type", result.GetObject().String()).
+				Msg("object type currently not supported")
+		}
+	}
+}
+
+func (s *Source) hasChanged(page *notion.Page) bool {
+	// see discussion in docs/cdc.md
+	lastTopMinute := time.Now().Truncate(time.Minute)
+	return page.LastEditedTime.After(s.lastEditedTime) &&
+		page.LastEditedTime.Before(lastTopMinute)
+}
+
+func (s *Source) getPages(ctx context.Context, cursor notion.Cursor) (*notion.SearchResponse, error) {
+	req := &notion.SearchRequest{
```

Comment:
Now, I know there's no other way to get a list of pages, but I have to call out the [docs](https://developers.notion.com/reference/post-search#limitations-of-search) which explicitly say:

> Search is not guaranteed to return everything, and the index may change as your integration is iterating through pages and databases resulting in unstable results.

This might be something to write into the known limitations.

Reply:
Given this info it seems that we should be prepared to fetch a block/page that doesn't exist and not fail because of that.

Reply:
Correct... And even if the search would be reliable, a page can be deleted between when we got its ID and when we attempted to fetch it. Thanks for calling it out, I'll implement this in a follow-up PR.

Reply:
Tracked here: #8.